src/app/libp2p_helper/src/libp2p_helper/bitswap_downloader.go

```diff
@@ -0,0 +1,333 @@
+		for root := range params {
+			malformed[root] = fmt.Errorf("Error reading block %s: %v", id, err)
+		}
+		return nil, malformed
+	}
+	children := make(map[BitswapBlockLink]map[root][]NodeIndex)
+	for root_, ixs := range params {
+		rp, hasRp := rootParams[root_]
+		if !hasRp {
+			bitswapLogger.Errorf("processBlock: didn't find root state for %s (root %s)",
+				id, codanet.BlockHashToCidSuffix(root_))
+			continue
+		}
+		schema := rp.getSchema()
+		hasRootIx := false
+		for _, ix := range ixs {
+			if ix == 0 {
+				hasRootIx = true
+				break
+			}
+		}
+		if hasRootIx {
+			blockData, dataLen, err := ExtractLengthFromRootBlockData(fullBlockData)
+			if err == nil && len(blockData) < 1 {
+				err = errors.New("error reading tag from block")
+			}
+			tag := rp.getTag()
+			if err == nil {
+				tag_ := BitswapDataTag(blockData[0])
+				if tag_ != tag {
+					err = fmt.Errorf("tag mismatch: %d != %d", tag_, tag)
+				}
+			}
+			if err == nil {
+				dataConf, hasDataConf := tagConfig[tag]
+				if !hasDataConf {
+					err = fmt.Errorf("no tag config for tag %d", tag)
+				} else if dataConf.maxSize < dataLen-1 {
+					err = fmt.Errorf("data is too large: %d > %d", dataLen-1, dataConf.maxSize)
+				}
+			}
+			if err != nil {
+				malformed[root_] = fmt.Errorf("error reading root block %s: %v", id, err)
+				continue
+			}
+			schema_ := MkBitswapBlockSchemaLengthPrefixed(maxBlockSize, dataLen)
+			schema = &schema_
+			rp.setSchema(schema)
+		}
+		if schema == nil {
+			bitswapLogger.Errorf("Invariant broken for %s (root %s): schema not set for non-root block",
+				id, codanet.BlockHashToCidSuffix(root_))
+			continue
+		}
+		for _, ix := range ixs {
+			if len(block.RawData()) != schema.BlockSize(ix) {
+				malformed[root_] = fmt.Errorf("unexpected size for block #%d (%s) of root %s: %d != %d",
+					ix, id, codanet.BlockHashToCidSuffix(root_), len(block.RawData()), schema.BlockSize(ix))
+				break
+			}
+			if len(links) != schema.LinkCount(ix) {
+				malformed[root_] = fmt.Errorf("unexpected link count for block %s of root %s: %d != %d (fullLinkBlocks: %d, ix: %d)",
+					id, codanet.BlockHashToCidSuffix(root_), len(links), schema.LinkCount(ix), schema.fullLinkBlocks, ix)
+				break
+			}
+			fstChildId := di.FirstChildId(ix)
+			for childIx, link := range links {
+				if children[link] == nil {
+					children[link] = make(map[root][]NodeIndex)
+				}
+				children[link][root_] = append(children[link][root_], fstChildId+NodeIndex(childIx))
+			}
+		}
+	}
+	return children, malformed
+}
+
+// processDownloadedBlock is a big-step transition of root block retrieval state machine
+// It transits state for a single block
+func processDownloadedBlock(block blocks.Block, bs BitswapState) {
+	bs.CheckInvariants()
+	id := block.Cid()
+	nodeDownloadParams := bs.NodeDownloadParams()
+	rootDownloadStates := bs.RootDownloadStates()
+	depthIndices := bs.DepthIndices()
+	oldPs, foundRoot := nodeDownloadParams[id]
+	delete(nodeDownloadParams, id)
```

Comment:
What happens if we a processing the same downloaded block in parallel? Is that possible? I'm worried about potential race conditions on accessing the node download params.

Reply:
You can check that `Loop` is only ran once. All these data structures are accessed only from functions called from within the `Loop()`. These are most importantly `kickStartRootDownload` and `processDownloadedBlock`. You can check that within these functions no new goroutines are produced.

So, basically, the execution is single-threaded by design.